util/kubeconfig/kubeconfig.go

```diff
@@ -174,3 +144,98 @@ func GenerateSecretWithOwner(clusterName client.ObjectKey, data []byte, owner me
 		},
 	}
 }
+
+// NeedsClientCertRotation returns whether any of the Kubeconfig secret's client certificates will expire before the given threshold.
+func NeedsClientCertRotation(configSecret *corev1.Secret, threshold time.Duration) (bool, error) {
+	now := time.Now()
+
+	data, err := toKubeconfigBytes(configSecret)
+	if err != nil {
+		return false, err
+	}
+
+	config, err := clientcmd.Load(data)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to convert kubeconfig Secret into a clientcmdapi.Config")
+	}
+
+	for _, authInfo := range config.AuthInfos {
+		cert, err := certs.DecodeCertPEM(authInfo.ClientCertificateData)
+		if err != nil {
+			return false, err
+		}
+		if cert.NotAfter.Sub(now) < threshold {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// UpdateSecret regenerates the Kubeconfig for the given secret.
+func UpdateSecret(ctx context.Context, c client.Client, configSecret *corev1.Secret) error {
```

Comment:
Naming and semantics feel a bit confusing here, would it make sense to having some type of composition of this with the `generateKubeConfig` function above to differentiate the action of generating the kubeconfig from the act of updating it on the server?

Alternatively, maybe it would make sense to use the same arguments as `CreateSecretWithOwner` here to avoid confusion around the passed in secret not actually being passed to the update call.

Reply:
Yeah, I wasn't totally happy with the objects/fns here too. What do you mean by
> avoid confusion around the passed in secret not actually being passed to the update call

configSecret is the arg, and is what gets passed to the Update call.

Reply:
Apologies for not being clearer. My confusion was stemming from the fact that we are modifying the `corev1.Secret` resource that is being passed in prior to calling `Update` rather than using the secret provided as-is to perform the `Update`

Reply:
I renamed it to `RegenerateSecret`, that seemed like the simplest thing to eliminate the confusion